core/controllers/reader.py

```diff
@@ -120,32 +132,89 @@ def classify(exp_id, state, answer, params):
         state.interaction.id)
     normalized_answer = interaction_instance.normalize_answer(answer)
 
-    # Find the first group that satisfactorily matches the given answer. This
-    # is done by ORing (maximizing) all truth values of all rules over all
-    # answer groups. The group with the highest truth value is considered the
-    # best match.
     best_matched_answer_group = None
     best_matched_answer_group_index = len(state.interaction.answer_groups)
+    best_matched_rule_spec = None
     best_matched_rule_spec_index = None
-    best_matched_truth_value = 0.0
+    best_matched_truth_value = rule_domain.CERTAIN_FALSE_VALUE
+    matched_rule_type = None
+
+    fs = fs_domain.AbstractFileSystem(fs_domain.ExplorationFileSystem(exp_id))
+    input_type = interaction_instance.answer_type
+
+    # Find the first hard rule that matches
     for (answer_group_index, answer_group) in enumerate(
             state.interaction.answer_groups):
-        fs = fs_domain.AbstractFileSystem(
-            fs_domain.ExplorationFileSystem(exp_id))
-        input_type = interaction_instance.answer_type
-        ored_truth_value = 0.0
-        best_rule_spec_index = None
-        for (rule_spec_index, rule_spec) in enumerate(answer_group.rule_specs):
-            evaluated_truth_value = rule_domain.evaluate_rule(
-                rule_spec, input_type, params, normalized_answer, fs)
-            if evaluated_truth_value > ored_truth_value:
-                ored_truth_value = evaluated_truth_value
-                best_rule_spec_index = rule_spec_index
-        if ored_truth_value > best_matched_truth_value:
+        ored_truth_value = rule_domain.CERTAIN_FALSE_VALUE
+        best_rule_spec = None
+        for (rule_spec_index, rule_spec) in enumerate(
+                answer_group.rule_specs):
+            if rule_spec.rule_type != rule_domain.FUZZY_RULE_TYPE:
+                evaluated_truth_value = rule_domain.evaluate_rule(
+                    rule_spec, input_type, params, normalized_answer, fs)
+                if evaluated_truth_value > ored_truth_value:
+                    ored_truth_value = evaluated_truth_value
+                    best_rule_spec_index = rule_spec_index
+                    best_rule_spec = rule_spec
+        if ored_truth_value == rule_domain.CERTAIN_TRUE_VALUE:
             best_matched_truth_value = ored_truth_value
             best_matched_answer_group = answer_group
             best_matched_answer_group_index = answer_group_index
             best_matched_rule_spec_index = best_rule_spec_index
+            matched_rule_type = 'hard'
```

Comment:
Sorry, I just noticed this. I don't think we need to expose this information to the caller. Ie, does this have any value for the learner or editor views? If it was done for tests, can't the rule_spec_index be used to validate whether it was a hard, soft, or classification rule, instead? I think rule_spec_index is guaranteed to be able to determine whether a rule was classified as hard or soft. If the rule is not within the training data of the soft rule, then it's guaranteed to be classifier.

To further tighten your tests, you could also swap out the string classifier function to verify it is called. This would be better than just assuming a soft rule is using the classifier, per above.